cpp/modmesh/toggle/RadixTree.hpp

```diff
@@ -179,5 +200,151 @@ class RadixTree
     key_type m_unique_id = 0;
 }; /* end class RadixTree */
 
+// The profiling result of the caller
+struct CallerProfile
+{
+    CallerProfile()
+        : total_time(0)
+    {
+    }
+
+    void start_stopwatch()
+    {
+        start_time = std::chrono::high_resolution_clock::now();
+        is_running = true;
+    }
+
+    void stop_stopwatch()
+    {
+        auto end_time = std::chrono::high_resolution_clock::now();
+        auto elapsed_time = std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start_time);
+        total_time += elapsed_time;
+        call_count++;
+    }
+
+    std::chrono::high_resolution_clock::time_point start_time;
+    std::function<void()> cancel_callback;
+    std::string caller_name;
+    std::chrono::nanoseconds total_time; /// use nanoseconds to have higher precision
+    int call_count = 0;
+    bool is_running = false;
+}; /* end struct CallerProfile */
+
+class CallProfilerTest; // for gtest
+
+/// The profiler that profiles the hierarchical caller stack.
+class CallProfiler
+{
+public:
+    /// A singleton.
+    static CallProfiler & instance()
+    {
+        static CallProfiler instance;
+        return instance;
+    }
+
+    CallProfiler(CallProfiler const &) = delete;
+    CallProfiler(CallProfiler &&) = delete;
+    CallProfiler & operator=(CallProfiler const &) = delete;
+    CallProfiler & operator=(CallProfiler &&) = delete;
+    ~CallProfiler() = default;
+
+    // Called when a function starts
+    void start_caller(const std::string & caller_name, std::function<void()> cancel_callback)
+    {
+        auto start_time = std::chrono::high_resolution_clock::now();
+        m_radix_tree.entry(caller_name);
+        CallerProfile & callProfile = m_radix_tree.get_current_node()->data();
+        callProfile.caller_name = caller_name;
+        callProfile.start_stopwatch();
+    }
+
+    // Called when a function ends
+    void end_caller(const std::string & caller_name)
+    {
+        CallerProfile & callProfile = m_radix_tree.get_current_node()->data();
+
```

Comment:
My personal preference is to avoid the unnecessary blank lines or even make them one-liner, e.g.,

```cpp
callProfile.stop_stopwatch(); // Update profiling information
m_radix_tree.move_current_to_parent(); // Pop the caller from the call stack
```

I feel it is more readable than

```cpp

// Update profiling information
callProfile.stop_stopwatch();

// Pop the caller from the call stack
m_radix_tree.move_current_to_parent();
```

But the latter is fine.  I don't have a problem with it.

Reply:
sure, let me change.

Reply:
I also prefer local variables to use `snake_case`.  But in the same way, I have not a problem with the style in the local scope.

Reply:
ops, I missed this one.